File: backend/src/main/java/info/gamewise/lor/videos/entity/VideoPredicate.java
package info.gamewise.lor.videos.entity;

import com.github.brunosc.lor.domain.LoRRegion;
import com.querydsl.core.BooleanBuilder;
import com.querydsl.core.types.Predicate;
import info.gamewise.lor.videos.port.in.GetVideosUseCase.SearchParams;
import org.springframework.data.domain.Sort;

import static java.util.Collections.singletonList;
import static org.springframework.util.CollectionUtils.isEmpty;

class VideoPredicate {

    private final SearchParams params;
    private final Sort sort;

    VideoPredicate(SearchParams params) {
        this.params = params;
        this.sort = sortByPublishedAt();
    }

    Predicate predicate() {
        QVideoJpaEntity q = new QVideoJpaEntity("videos");
        BooleanBuilder predicate = new BooleanBuilder();

        if (!isEmpty(params.regions())) {
            for (String region : params.regions()) {
                predicate.and(q.regions.contains(LoRRegion.fromCode(region)));
            }
        }

        if (!isEmpty(params.champions())) {
            for (String champion : params.champions()) {
                predicate.and(q.champions.contains(champion));
            }
        }

        if (!isEmpty(params.channels())) {
            predicate.and(q.channel.in(params.channels()));
        }

        return predicate;
    }

    Sort sort() {
        return this.sort;
    }

    private Sort sortByPublishedAt() {
        Sort.Order sortOrderPublishedAt = new Sort.Order(Sort.Direction.DESC, "publishedAt").ignoreCase();
        return Sort.by(singletonList(sortOrderPublishedAt));
    }
}

File: backend/src/main/kotlin/info/gamewise/lor/videos/entity/VideoPredicate.kt
package info.gamewise.lor.videos.entity

import info.gamewise.lor.videos.port.`in`.SearchParams
import org.springframework.data.domain.Sort

internal class VideoPredicate(params: SearchParams) {

    private val params: SearchParams
    private val sort: Sort

    init {
        this.params = params
        sort = sortByPublishedAt()
    }

    fun sort(): Sort {
        return sort
    }

    private fun sortByPublishedAt(): Sort {
        val sortOrderPublishedAt = Sort.Order(Sort.Direction.DESC, "publishedAt")
        return Sort.by(listOf(sortOrderPublishedAt))
    }
}